src/qiskit_qasm3_import/expression.py

```diff
@@ -47,6 +48,19 @@ def join_integer_types(left: _IntegerT, right: _IntegerT) -> _IntegerT:
     return types.Int(const, size)
 
 
+def join_numeric_types(left: _NumericT, right: _NumericT) -> _NumericT:
+    const = left.const and right.const
+    if isinstance(left, types.Float) and isinstance(right, types.Float):
+        size = None if left.size is None or right.size is None else max((left.size, right.size))
+        return types.Float(const, size)
+    if isinstance(left, types.Float):
+        return types.Float(const, left.size)
+    if isinstance(right, types.Float):
+        return types.Float(const, right.size)
+    # Cast because this can't return the never type since neither of the inputs are never.
+    return cast(_NumericT, join_integer_types(left, right))
```

Comment:
This is just for static type checking, right? And in particular just for `mypy`, which is apparently not clever enough to infer that the return type is `_NumericT`. Why are we trying to satisfy `mypy` if we are not using it in testing? I think it's difficult to remain compliant without testing

Reply:
Oh yeah, it's mostly because I forgot that we _weren't_ using mypy - I have it locally hooked into my editor and it was moaning about this line.

Reply:
I can run `mypy` locally as part of developing. We could add it local `tox` config. Or even CI. What do you think? I have some experience with it. My feelings flip back and forth between, this is kind of useful, to this is a major PITA.

Reply:
Maybe let's leave everything as it is right now, and I'll look into it whether it's worth it to enable properly and add to CI.  I don't think this package is 100% typed as things stand already, but I don't exactly remember.